Skew join optimizer

damnMeddlingKid opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Overview
This PR introduces a new optimization rule to handle joins where keys have a skewed distribution (some keys are much more frequent than others) using the salted join technique.

<img width="727" alt="image" src="https://user-images.githubusercontent.com/422763/181656656-5f23c57b-5428-476d-a0cc-f0635e1cd4b4.png">


here are the steps for breaking up skew in the optimizer:

- User provides the name of the skewed column that they want to join on and the values of keys that are skewed.
- On the left side a new column is created called `randPart`, for the skewed rows `randPart=random(N)` where N is the number of buckets to partition skew. for non-skewed rows `randPart=0` 
- On the right side we create N copies of rows with skewed values and assign them a `skewPart` of 0,1,..N respectively
- Add an additional condition to the join clause to join on `left.randPart = right.skewPart`
- Drop `randPart` column after the join 

see https://github.com/trinodb/trino/issues/1261 for further discussion

## Assumptions 

- The rule only applies to inner joins and left joins, other join types will have errors because of the duplicate rows on the right side.
- I'm only considering partitioned joins because there won't be much benefit for replicated joins as they are fast enough.
- For now i've made the decision to not handle the case where the skew is on the right side. but i think in the future we can generalize the optimizer to handle skew in either side.
- The optimizer traverses the plan to find all symbols that reference skewed symbols and propagates this information through nodes. it does not propagate skewed symbols through JoinNodes, so a join on a skewed join will not be rewritten even if it references a skewed symbol. Nested joins are currently out of scope because it is possible for joins to alter the distribution of skewness. However nesting joins against a skewed fact table is a popular pattern in queries. We should tackle this in future extensions to this work.


## Details of the changes

**SystemSessionProperties.java** 

Introduces two new session variables

- SKEWED_JOIN_METADATA: This is an array of arrays of the form Array[Array[tablename, skewedcolumn, skewedvalue, skewedvalue]]
- SKEWED_JOIN_REPLICATION_FACTOR: The number of buckets to partition skewed rows by, defaults to 3.

**SkewJoinOptimizer.java**

- SkewJoinConfig: Parses the session variables for and extracts the table, column and skewed values.
- PlanRewriter: Traverses the plan and finds all skewed symbols that might reference the user specified skewed column. When it finds a join that uses a skewed symbol it rewrites it.
- JoinRewriter: Rewrites skewed joins with a new join that has its left side projected with a random symbol and the right side unnested to produce duplicated rows, also modifies the join clause to include the random symbols.

## Future work
- Use a hint framework to specify the configuration rather than session variables, that way we don't need to traverse the plan to find join nodes and symbols.
- Handle skew on the right side as well.
- Handle nested joins.
- We can extend a similar pattern to handle range joins, im going to be exploring this angle.

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

related to: https://github.com/trinodb/trino/issues/1261

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Introduced a skew join optimizer. ({issue} https://github.com/trinodb/trino/issues/1261)
```

cc: @amoghmargoor

damnMeddlingKid commented:
Signed the CLA

hashhar commented:
FYI @martint @sopel39

findepi commented:
cc @raunaqmorarka @skrzypo987

sopel39 commented:
Skewed columns need to be derived from stats, so there should be no requirement for  `SKEWED_JOIN_METADATA`

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+    }
+}
+
+public class SkewJoinOptimizer
```
Could you make it an iterative optimizer rule?

damnMeddlingKid replied:
I don't think i can make this a rule yet, i need to be able to traverse the graph and collect symbols that might be skewed.

sopel39 replied:
> I don't think i can make this a rule yet, i need to be able to traverse the graph and collect symbols that might be skewed.

You could use `StatsCalculator` from a rule to identify how many distinct values given join symbols has

damnMeddlingKid replied:
didn't know about that. I'm not sure how it would work though, i need to know that the symbol is the same as the column the user has specified.

sopel39 replied:
> i need to know that the symbol is the same as the column the user has specified.

You don't need user input.

You can use `io.trino.cost.StatsCalculator#calculateStats` to get statistics of join source nodes. Then you can check `io.trino.cost.SymbolStatsEstimate#getDistinctValuesCount` for join key symbol.
If `SymbolStatsEstimate#getDistinctValuesCount` is lower than some threshold (e.g. `number of nodes * task_concurrency`) then you could trigger your rule.

sopel39 replied:
> didn't know about that. I'm not sure how it would work though, i need to know that the symbol is the same as the column the user has specified.

You are right. You need actual skewed values which CBO cannot provide. @martint are we fine adding session property like `SKEWED_JOIN_METADATA` of we would like to have something more adaptive?

sopel39 commented:
Thanks @damnMeddlingKid. I've made some initial comments

damnMeddlingKid commented:
> Skewed columns need to be derived from stats, so there should be no requirement for `SKEWED_JOIN_METADATA`
From the original issue. @martint suggested getting the feature in with a session property so users can start using the functionality manually and the over time we can build on top of it to improve the user experience by using stats. https://github.com/trinodb/trino/issues/1261#issuecomment-524572774

I think its a good approach, we can build incrementally from here.

hashhar commented:
@damnMeddlingKid A session property is good to have but unless the implementation has issues which make it not ideal to be used generally I suggest we enable it by default and have the session property as a kill switch if needed. People very rarely enable features explicitly (and this also matches what we have done for other improvements in the past).

martint commented:
@hashhar, it can't be enabled by default because it requires information that is not available otherwise (e.g., which columns are skewed, what values, etc). Until such information can be provided by table metadata, a session property is a reasonable approach.

sopel39 commented:
@damnMeddlingKid Can you use broadcast join for your skewed queries or build side is too big?

damnMeddlingKid commented:
> @damnMeddlingKid Can you use broadcast join for your skewed queries or build side is too big?
We are still working on collecting data from our workload to isolate skewed joins, We won't be able to use replicated joins in all cases. This rule doesn't apply if the join is replicated.

damnMeddlingKid commented:
Wanted to bump this, would love to get people's thoughts on the change.

damnMeddlingKid commented:
cc: @sopel39 @raunaqmorarka

damnMeddlingKid commented:
Hey folks, was wondering how to proceed here.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+
+    public SkewJoinOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
```
```suggestion
        this.metadata = requireNonNull(metadata, "metadata is null");
```

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+public class SkewJoinOptimizer
+        implements PlanOptimizer
+{
+    Metadata metadata;
```
```suggestion
    private final Metadata metadata;
```

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        return plan.accept(new PlanRewriter(session, metadata, skewedTables, rewriter), new HashMap<>());
+    }
+
+    static class SkewedContext
```
```suggestion
    private static class SkewedContext
```

martint replied:
This could be a `record`

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        }
+    }
+
+    static class PlanRewriter
```
```suggestion
    private static class PlanRewriter
```

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+
+        public PlanRewriter(Session session, Metadata metadata, Map<String, SkewJoinConfig> skewedTables, JoinRewriter rewriter)
+        {
+            this.session = session;
+            this.metadata = metadata;
+            this.skewedTables = skewedTables;
+            this.joinRewriter = rewriter;
```
Null checks

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
```
This is not correct. ExchangeNode remaps input symbols to output symbols.

However, since this optimizer needs to run before exchanges are added. I'd remove this and just fail with an error in case this node is encountered.

damnMeddlingKid replied:
good point, will fail here.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        @Override
+        public PlanNode visitProject(ProjectNode node, Map<Symbol, SkewedContext> context)
+        {
+            /*
+                If any symbol assignments refers to a skewed symbol then add that symbol to the set of skewed symbols.
+            * */
```
```suggestion
            // If any symbol assignments refers to a skewed symbol then add that symbol to the set of skewed symbols.
```

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+                        Expression value = entry.getValue();
+                        if (entry.getValue() instanceof SymbolReference) {
+                            Symbol skewedSymbol = Symbol.from(value);
+                            if (context.containsKey(Symbol.from(value)) && !context.containsKey(entry.getKey())) {
```
This context cannot be global to the whole plan. It is technically possible for symbol names to get reused across unrelated plan nodes.

damnMeddlingKid replied:
Interesting!, i didn't know that. The context here is created independently for each child of a join node (left/right), and we don't allow nested joins. Does that still violate the assumption ?.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, Map<Symbol, SkewedContext> context)
+        {
+            final String qualifiedTableName = metadata.getTableMetadata(session, node.getTable()).getQualifiedName().toString();
```
Don't use final for local variables

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+                        .map(entry -> new Pair<>(entry.getKey(), metadata.getColumnMetadata(session, node.getTable(), entry.getValue())))
+                        .filter(entry -> entry.getB().getName().equals(skewConfig.getColumnName()))
+                        .findFirst();
+                symbolAndMetadata.ifPresent(m -> {
```
Avoid abbreviations

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        }
+    }
+
+    static class SkewJoinConfig
```
```suggestion
    private static class SkewJoinConfig
```

martint replied:
This class can be a `record`

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+         * Parses and validates the user supplied session data.
+         * */
+
+        final String tableName;
+        final String columnName;
+        final List<String> values;
```
private

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        }
+    }
+
+    static class JoinRewriter
```
```suggestion
    private static class JoinRewriter
```

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        {
+            final String qualifiedTableName = metadata.getTableMetadata(session, node.getTable()).getQualifiedName().toString();
+
+            if (this.skewedTables.containsKey(qualifiedTableName)) {
```
This doesn't handle case sensitivity issues or quoted identifiers (e.g., what happens if the table name contains a special character?)

damnMeddlingKid replied:
I've switched this to match on lower case, but im not sure how to handle the spaces and escape characters. We would need a utility to normalize table names ?.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.StringLiteral;
+import io.trino.sql.tree.SymbolReference;
+import oshi.util.tuples.Pair;
```
Spurious import. Instead of using a generic `Pair`, create a dedicated `record` that contains the information needed.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        {
+            final String qualifiedTableName = metadata.getTableMetadata(session, node.getTable()).getQualifiedName().toString();
+
+            if (this.skewedTables.containsKey(qualifiedTableName)) {
```
`this.` qualifier not needed.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+
+    static class SkewJoinConfig
+    {
+        /*
+         * Parses and validates the user supplied session data.
+         * */
```
Move this to a comment on the class:

```
// Parses and validates ...
private static class SkewJoinConfig {
    ...
}

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+            return values;
+        }
+
+        static Optional<SkewJoinConfig> getMetadata(List<String> rawMetadata)
```
This is part of this class' public interface, so mark it `public`. I'd also rename it to `valueOf` or `newInstance` to indicate its purpose is to create an instance of the class.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+
+    static class JoinRewriter
+    {
+        private static final LongLiteral DEFAULT_PROJECT_VALUE = new LongLiteral("0");
```
Use `GenericLiteral(...)`. The type of a LongLiteral depends on the value it contains (a side effect of the AST and IR using the same expression hierarchy)

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+    static class JoinRewriter
+    {
+        private static final LongLiteral DEFAULT_PROJECT_VALUE = new LongLiteral("0");
+        private static final ArrayConstructor DEFAULT_UNNEST_VALUE = new ArrayConstructor(ImmutableList.of(DEFAULT_PROJECT_VALUE));
```
We're trying to get rid of ArrayConstructor from the IR (see https://github.com/trinodb/trino/pull/14415/files). Replace this with an invocation of the internal `$array` function.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+        private static final LongLiteral DEFAULT_PROJECT_VALUE = new LongLiteral("0");
+        private static final ArrayConstructor DEFAULT_UNNEST_VALUE = new ArrayConstructor(ImmutableList.of(DEFAULT_PROJECT_VALUE));
+
+        private final LongLiteral maxReplicationFactor;
+        private final ArrayConstructor arrayPartitioner;
```
See above comments

martint commented on core/trino-main/src/test/java/io/trino/sql/planner/assertions/ExpressionVerifier.java:
```diff
@@ -208,6 +210,23 @@ private static String getValueFromLiteral(Node expression)
     }
 
+    @Override
+    protected Boolean visitArrayConstructor(ArrayConstructor actual, Node expectedExpression)
```
See comment above

martint commented on testing/trino-tests/src/test/java/io/trino/tests/TestSkewJoinQueries.java:
```diff
@@ -0,0 +1,66 @@
+    @Test
+    public void testSkewedLeftJoin()
+    {
+        String sql = "SELECT o.orderkey, SUM(l.suppkey) FROM orders o LEFT JOIN lineitem l ON o.orderkey = l.orderkey"
+                + " WHERE l.suppkey > 10 GROUP BY o.orderkey ORDER BY 2 DESC LIMIT 10";
```
Use text blocks:

```java
String sql = """
        SELECT o.orderkey, SUM(l.suppkey)
        FROM orders o LEFT JOIN lineitem l ON o.orderkey = l.orderkey"
        WHERE l.suppkey > 10
        GROUP BY o.orderkey
        ORDER BY 2 DESC
        LIMIT 10
        """;
```

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+
+        private SkewJoinConfig(List<String> rawMetadata)
+        {
+            tableName = rawMetadata.get(TABLE_INDEX);
+            columnName = rawMetadata.get(COLUMN_INDEX);
+            values = ImmutableList.copyOf(rawMetadata.subList(VALUE_INDEX, rawMetadata.size()));
```
This should be done in the `getMetadata` method below, and the object should be constructed with the constituent parts directly.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+            return columnName;
+        }
+
+        public List<String> getValues()
```
Where is this used? I don't see any references in this PR.

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java:
```diff
@@ -0,0 +1,440 @@
+                return joinNode;
+            }
+
+            final SymbolReference skewedLeftKey = skewedClause.getLeft().toSymbolReference();
+            final SymbolReference skewedRightKey = skewedClause.getRight().toSymbolReference();
+            final List<Expression> skewedExpressionValues = context.values.stream()
```
Don't use final for local variables

martint commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -59,6 +61,8 @@
     public static final String JOIN_MAX_BROADCAST_TABLE_SIZE = "join_max_broadcast_table_size";
     public static final String JOIN_MULTI_CLAUSE_INDEPENDENCE_FACTOR = "join_multi_clause_independence_factor";
     public static final String DISTRIBUTED_INDEX_JOIN = "distributed_index_join";
+    public static final String SKEWED_JOIN_METADATA = "skewed_join_metadata";
```
Instead of `skewed_join_metadata`, this should be just `skewed_values` or something similar. There's nothing specific to joins in this option. In fact, it could be used for other operations such as aggregations, for instance.

damnMeddlingKid replied:
good point, i think aggregations and range joins could benefit.

damnMeddlingKid commented:
I've removed the array constructor and fixed all the scope issues. Will look into using the Record class and fixing getMetadata this week.

damnMeddlingKid commented:
> We're gradually trying to get rid of whole-plan optimizers based on PlanOptimizer in favor of incremental optimizers based on Rule. Unfortunately, this specific optimization requires propagating "traits" up the the plan, but we don't yet have a mechanism to do that in the Rule-based optimizer framework.

I think thats a good idea. We have a PR that introduces a hint framework, if we get that in then users can configure this optimization directly on the join/select that is skewed. That will make it possible for the optimizer to become a rule. Should we hold off on this till then ?.

> The skewed values in the session property are encoded are "stringly-typed". This could be a source of subtle errors and would require specific knowledge about how such strings get converted to the required type at runtime.

True, i've been relying on infering the type from the source symbols and then casting. this should work for the common key types, i could maybe also allow the user to pass a sql type. what do you think ?.

damnMeddlingKid commented:
I've addressed all remaining comments and have two open comments. @martint could you take a look.

I'll also try and run some tests on data this week and pull some numbers.

amoghmargoor commented:
This will help in making this into iterative optimiser rule - https://github.com/trinodb/trino/issues/9498 and also remove session property usage for specifying skewed values. However even in that case there are chances of traversing same plan subtrees multiple times for different invocation of rules - provided hints are provided for more than one join in query.

mosabua commented:
:wave: @damnMeddlingKid @martint  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

mosabua commented:
@martint do we need to revive this effort or close the PR
